txpipe/covariance.py

```diff
@@ -61,206 +77,419 @@ def read_cosmology(self):
+                cache_key1 = (tracer_comb1[i], tracer_comb2[j])
+                cache_key2 = (tracer_comb2[j], tracer_comb1[i])
+                if cache_key1 in cache:
+                    cl[local_key] = cache[cache_key1]
+                elif cache_key2 in cache:
+                    cl[local_key] = cache[cache_key2]
                 else:
-                    bias = (tracer1.z, np.ones_like(tracer1.z))
-                    cTracer1 = ccl.NumberCountsTracer(cosmo, has_rsd=False, bias=bias, dndz=dndz1)
-                    warnings.warn("Not including bias in fiducial cosmology")
+                    # If not cached then we must compute
+                    t1 = tracer_comb1[i]
+                    t2 = tracer_comb2[j]
+                    c = ccl.angular_cl(cosmo, ccl_tracers[t1], ccl_tracers[t2], ell)
+                    print("Computed C_ell for ", cache_key1)
+                    cache[cache_key1] = c
+                    cl[local_key] = c
+
+        # The shape noise C_ell values.
+        # These are zero for cross bins and as computed earlier for auto bins
+        SN={}
+        SN[13] = tracer_Noise[tracer_comb1[0]] if tracer_comb1[0] == tracer_comb2[0] else 0
+        SN[24] = tracer_Noise[tracer_comb1[1]] if tracer_comb1[1] == tracer_comb2[1] else 0
+        SN[14] = tracer_Noise[tracer_comb1[0]] if tracer_comb1[0] == tracer_comb2[1] else 0
+        SN[23] = tracer_Noise[tracer_comb1[1]] if tracer_comb1[1] == tracer_comb2[0] else 0
+
+
+        # The overall normalization factor at the front of the matrix
+        if self.do_xi:
+            norm = np.pi * 4 * meta['f_sky']
+        else: 
+            norm = (2*ell + 1) * np.gradient(ell) * meta['f_sky']
+
+        # The coupling is an identity matrix at least when we neglect
+        # the mask
+        coupling_mat = {}
+        coupling_mat[1324] = np.eye(len(ell))
+        coupling_mat[1423] = np.eye(len(ell))
+
+        # Initial covariance of C_ell components
+        cov = {}
+        cov[1324] = np.outer(cl[13] + SN[13], cl[24] + SN[24]) * coupling_mat[1324]
+        cov[1423] = np.outer(cl[14] + SN[14], cl[23] + SN[23]) * coupling_mat[1423]
+
+        # for shear-shear components we also add a B-mode contribution
+        first_is_shear_shear = ('source' in tracer_comb1[0]) and ('source' in tracer_comb1[1])
+        second_is_shear_shear = ('source' in tracer_comb2[0]) and ('source' in tracer_comb2[1])
+
+        if self.do_xi and (first_is_shear_shear or second_is_shear_shear):
+            # this adds the B-mode shape noise contribution.
+            # We assume B-mode power (C_ell) is 0
+            Bmode_F = 1
+            if xi_plus_minus1 != xi_plus_minus2:
+                # in the cross term, this contribution is subtracted.
+                # eq. 29-31 of https://arxiv.org/pdf/0708.0387.pdf
+                Bmode_F=-1 
+            cov[1324] += np.outer(cl[13]*0 + SN[13], cl[24]*0 + SN[24]) * coupling_mat[1324] * Bmode_F
```

Comment:
Are the components here that are multiplied by zero needed to maintain the right shape? (I assume they representing are B-mode signal C_ells).  If that's right could you add a comment to that effect?

Reply:
yep